src/Catalyst.Core.Modules.Dfs/CoreApi/UnixFsApi.cs

```diff
@@ -0,0 +1,467 @@
+            {
+                using (var archive = TarArchive.CreateOutputTarArchive(tarStream))
+                {
+                    archive.IsStreamOwner = false;
+                    await AddTarNodeAsync(cid, cid.Encode(), tarStream, cancel).ConfigureAwait(false);
+                }
+            }
+
+            ms.Position = 0;
+            return ms;
+        }
+
+        async Task AddTarNodeAsync(Cid cid, string name, TarOutputStream tar, CancellationToken cancel)
+        {
+            var block = await _blockApi.GetAsync(cid, cancel).ConfigureAwait(false);
+            var dm = new DataMessage
+            {
+                Type = DataType.Raw
+            };
+            DagNode dag = null;
+
+            if (cid.ContentType == "dag-pb")
+            {
+                dag = new DagNode(block.DataStream);
+                dm = Serializer.Deserialize<DataMessage>(dag.DataStream);
+            }
+
+            var entry = new TarEntry(new TarHeader());
+            var header = entry.TarHeader;
+            header.Mode = 0x1ff; // 777 in octal
+            header.LinkName = String.Empty;
+            header.UserName = String.Empty;
+            header.GroupName = String.Empty;
+            header.Version = "00";
+            header.Name = name;
+            header.DevMajor = 0;
+            header.DevMinor = 0;
+            header.UserId = 0;
+            header.GroupId = 0;
+            header.ModTime = DateTime.Now;
+
+            if (dm.Type == DataType.Directory)
+            {
+                header.TypeFlag = TarHeader.LF_DIR;
+                header.Size = 0;
+                tar.PutNextEntry(entry);
+                tar.CloseEntry();
+            }
+            else // Must be a file
+            {
+                var content = await ReadFileAsync(cid, cancel).ConfigureAwait(false);
+                header.TypeFlag = TarHeader.LF_NORMAL;
+                header.Size = content.Length;
+                tar.PutNextEntry(entry);
+                await content.CopyToAsync(tar, cancel);
+                tar.CloseEntry();
+            }
+
+            // Recurse over files and subdirectories
+            if (dm.Type == DataType.Directory)
+            {
+                foreach (var link in dag.Links)
+                {
+                    await AddTarNodeAsync(link.Id, $"{name}/{link.Name}", tar, cancel).ConfigureAwait(false);
+                }
+            }
+        }
+
+        IBlockApi GetBlockService(AddFileOptions options)
+        {
+            return options.OnlyHash
+                ? new HashOnlyBlockService()
+                : _blockApi;
+        }
+
+        /// <summary>
+        ///   A Block service that only computes the block's hash.
+        /// </summary>
+        class HashOnlyBlockService : IBlockApi
+        {
+            public IPinApi PinApi { get; set; }
+
+            public Task<IDataBlock> GetAsync(Cid id, CancellationToken cancel = default)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Remove the unused private method 'GetAsync'.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)